flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateCheckpointWriterTest.java

```diff
@@ -0,0 +1,133 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint.channel;
+
+import org.apache.flink.core.memory.HeapMemorySegment;
+import org.apache.flink.runtime.checkpoint.channel.ChannelStateWriter.ChannelStateWriteResult;
+import org.apache.flink.runtime.io.network.buffer.FreeingBufferRecycler;
+import org.apache.flink.runtime.io.network.buffer.NetworkBuffer;
+import org.apache.flink.runtime.state.InputChannelStateHandle;
+import org.apache.flink.runtime.state.memory.MemCheckpointStreamFactory.MemoryCheckpointOutputStream;
+import org.apache.flink.util.function.RunnableWithException;
+
+import org.junit.Test;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * {@link ChannelStateCheckpointWriter} test.
+ */
+public class ChannelStateCheckpointWriterTest {
+	private static final RunnableWithException NO_OP_RUNNABLE = () -> {
+	};
+	private final Random random = new Random();
+
+	@Test
+	public void testRecyclingBuffers() throws Exception {
+		ChannelStateCheckpointWriter writer = createWriter(new ChannelStateWriteResult());
+		NetworkBuffer buffer = new NetworkBuffer(HeapMemorySegment.FACTORY.allocateUnpooledSegment(10, null), FreeingBufferRecycler.INSTANCE);
+		writer.writeInput(new InputChannelInfo(1, 2), buffer);
+		assertTrue(buffer.isRecycled());
+	}
+
+	@Test
+	public void testFlush() throws Exception {
+		class FlushRecorder extends DataOutputStream {
+			private boolean flushed = false;
+
+			private FlushRecorder() {
+				super(new ByteArrayOutputStream());
+			}
+
+			@Override
+			public void flush() throws IOException {
+				flushed = true;
+				super.flush();
+			}
+		}
+
+		FlushRecorder dataStream = new FlushRecorder();
+		final ChannelStateCheckpointWriter writer = new ChannelStateCheckpointWriter(
+			1L,
+			new ChannelStateWriteResult(),
+			new ChannelStateSerializerImpl(),
+			NO_OP_RUNNABLE,
+			new MemoryCheckpointOutputStream(42),
+			dataStream
+		);
+
+		writer.completeInput();
+		writer.completeOutput();
+
+		assertTrue(dataStream.flushed);
```

Comment:
it is better to also verify the `ChannelStateWriteResult#isDone` when complete both input and output.

Reply:
There are already tests that check that result is completed.
However, they are on a higher level, so I added a separate test here too.